### DISCUSSION TITLE
Make empty changeset updates using `.optional()` , either ignore or bypass errors `QueryBuilderError` s.

### DISCUSSION THREAD

**Discussion by dessalines** [diesel-rs/diesel#4018] [Ideas]
### Versions

- **Rust:** - rustc 1.80.0-nightly (9c9b56879 2024-05-05)
- **Diesel:** - 2.1.6
- **Database:** - Postgres 16
- **Operating System** - Arch

### Feature Flags

- **diesel:** NA

## Problem Description

Oftentimes, when doing DB updates, we need to build changeset objects, where we don't know beforehand if its an empty changeset, IE if all the updated fields are `None`. 

When that's the case, diesel throws a runtime [QueryBuilderError](https://docs.rs/diesel/latest/diesel/result/enum.Error.html#variant.QueryBuilderError), which needs to be explicitly caught.

We'd like to either ignore, or bypass that error completely, rather than having to catch runtime QueryBuilderError s for every single update in our application.

The best solution IMO, would be to add this functionality to the existing `.optional()` , which already does a similar function: turning empty `get_result` from an error to an option. 

Adding this would also turn an empty update / changeset into an Option. IE, if `.optional()` is done, convert a QueryBuilderError into `None`, rather than throwing a runtime error.

Context: https://github.com/LemmyNet/lemmy/pull/4705

Also #598 , #1996

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) and the [discussion forum](https://github.com/diesel-rs/diesel/discussions) for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment]
@dessalines I've converted this to a discussion as this is strictly speaking a new feature. 
That written: I'm open to accept such an extension as long as it is reasonable named, documented, tested and only matches that single error case (and nothing else). 

The `.optional()` method is just provided by an extension trait implemented on [Result](https://docs.diesel.rs/2.1.x/src/diesel/result.rs.html#283-291) that internally matches the error and converts the result. That's something that could also be done in any other crate, so that's not even blocked having it in diesel  itself.

**Reply by dessalines** [reply]
Ignore my last comment, I figured out how to do this with the existing optional.

**Reply by dessalines** [reply]
k, [PR here](https://github.com/diesel-rs/diesel/pull/4019).
